Cargill/splinter
# Oauth stabilization fixes

# Testing

## Setup

1. Make sure your environment is setup to run splinterd locally:

```bash
$ cargo build --manifest-path splinterd/Cargo.toml --features experimental
$ cargo build --manifest-path cli/Cargo.toml --features experimental
$ export PATH=./target/debug:$PATH
$ export SPLINTER_HOME=/tmp/splinter
$ splinter database migrate
$ splinter cert generate
```

## Testing auth params

2. Run splinterd with Google as auth provider but configure it as a generic OpenID provider with additional params:

```bash
$ splinterd \
  --rest-api-endpoint http://localhost:8080 \
  --tls-insecure \
  --oauth-provider openid \
  --oauth-client-id <client-id> \
  --oauth-client-secret <client-secret> \
  --oauth-redirect-url http://localhost:8080/oauth/callback \
  --oauth-openid-url https://accounts.google.com/.well-known/openid-configuration \
  --oauth-openid-auth-param access_type=offline \
  --oauth-openid-auth-param prompt=consent \
  -vv
```

3. Go to http://localhost:8080/oauth/login?redirect_url=redirect in your browser and login (you will eventually be redirected and get a 404)

4. Verify that the new session has a refresh token (2nd to last value), which indicates that the auth params were added:

```bash
$ sqlite3 /tmp/splinter/data/splinter_state.db
sqlite> SELECT * FROM oauth_user_sessions;
o6AQzhOHlV7yMqrjpTiTgiDPcicwXjIQ|116381315131550692628|ya29.a0AfH6SMD7X8kpebidwV1Ql_Hn4YKim8MxNRhJIGrMBDNpMn4MJT7x77Y326sOkrTq_6_lfnqEzV7ld8V7qiB-O34nfLESMuZWAIf8M2m_zYkCNpXOSczfZz7TZGqDLxT8EDGsUE66j5k5HbZ9cjxTx4U8AAqfe2aZUS0ReqhAoqY|1//04iI1Qw-AR0obCgYIARAAGAQSNwF-L9IrkZOnArAW03diII5zDVguqdEh4HaELG6gJr9zLgomnEMGB0Ry95wAIDtRFsnvSFAPVvQ|1611597844
```

### Testing scopes

5. Run splinterd with Azure as auth provider but configure it as a generic OpenID provider with additional scopes:

```bash
$ splinterd \
  --rest-api-endpoint http://localhost:8080 \
  --tls-insecure \
  --oauth-provider openid \
  --oauth-client-id <client-id> \
  --oauth-client-secret <client-secret> \
  --oauth-redirect-url http://localhost:8080/oauth/callback \
  --oauth-openid-url https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration \
  --oauth-openid-scope offline_access \
  -vv
```

6. Repeat steps 3-4 above to verify the scope was added
